The last homework of the security-related functionality.

ipantazi opened the pull request:
Added corresponding properties (jwt.secret and jwt.expiration) to the application.properties files in the src/test/resources and src/main/resources folders.

Added method login() to the AuthenticationController class. 

Added class AuthenticationException and added it to CustomGlobalExceptionHandler class.

Added AuthenticationService, JwtAuthenticationFilter, JwtUtil, and OpenApiConfig classes.

 Added UserLoginRequestDto and UserLoginResponseDto records. 

Added AuthenticationManager method and jwtAuthenticationFilter to SecurityConfig class. 

Added jjwt dependency to pom.xml.

ipantazi committed:
Added corresponding properties (jwt.secret and jwt.expiration) to the application.properties files in the src/test/resources and src/main/resources folders.
Added method login() to the AuthenticationController class.
Added class AuthenticationException and added it to CustomGlobalExceptionHandler class.
Added AuthenticationService, JwtAuthenticationFilter, JwtUtil, and OpenApiConfig classes.
Added UserLoginRequestDto and UserLoginResponseDto records.
Added AuthenticationManager method and jwtAuthenticationFilter to SecurityConfig class.
Added jjwt dependency to pom.xml.

Elena-Bruyako commented on src/main/java/onlinebookstore/config/OpenApiConfig.java:
```diff
@@ -0,0 +1,22 @@
+    @Bean
+    public OpenAPI customOpenApi() {
+        return new OpenAPI()
+                .components(new Components().addSecuritySchemes("BearerAuth",
```
create constants for all strings

Elena-Bruyako commented on src/main/java/onlinebookstore/dto/user/UserLoginRequestDto.java:
```diff
@@ -0,0 +1,15 @@
+public record UserLoginRequestDto(
+        @Email(message = "Invalid format email.")
+        @Size(max = 50, message = "Email address should be exceed 50 characters.")
+        String email,
```
add `@NotBlank`

Elena-Bruyako commented on src/main/java/onlinebookstore/security/AuthenticationService.java:
```diff
@@ -0,0 +1,30 @@
+    private final JwtUtil jwtUtil;
+
+    public UserLoginResponseDto authenticate(UserLoginRequestDto requestDto) {
+        User userFromDB = userRepository.findByEmail(requestDto.email())
+                .orElseThrow(() -> new AuthenticationException("Invalid email or password."));
+
+        if (!passwordEncoder.matches(requestDto.password(), userFromDB.getPassword())) {
+            throw new AuthenticationException("Invalid email or password.");
+        }
```
```suggestion
        final Authentication authentication = manager
                .authenticate(
                        new UsernamePasswordAuthenticationToken(
                                requestDto.email(), requestDto.password()));
        String token = jwtUtil.generateToken(authentication.getName());
```

Elena-Bruyako commented on src/main/java/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
```
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

Elena-Bruyako commented on src/main/java/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,50 @@
+
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
```
create constant for "Bearer "

Elena-Bruyako commented on src/main/java/onlinebookstore/security/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,50 @@
+    private String getToken(HttpServletRequest request) {
+        String bearerToken = request.getHeader("Authorization");
+        if (StringUtils.hasText(bearerToken) && bearerToken.startsWith("Bearer ")) {
+            return bearerToken.substring(7);
```
```suggestion
            return bearerToken.substring(constant.length());
```

Elena-Bruyako reviewed (changes requested):
see comments

ipantazi committed:
Use Spring Security AuthenticationManager instead of manual password checking in AuthenticationService class.
Created constants for all strings in OpenApiConfig class.
Added @NotBlank in UserLoginRequestDto.
Added try-catch to JwtAuthenticationFilter and send JSON response with error.
